engine/pkg/externalresource/manager/gc_coordinator.go

```diff
@@ -0,0 +1,239 @@
+) (*notifier.Receiver[JobStatusChangeEvent], *notifier.Receiver[model.ExecutorStatusChange], error) {
+	jobSnapshot, jobWatchCh, err := c.jobInfos.WatchJobStatuses(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	executorSnapshot, executorReceiver, err := c.executorInfos.WatchExecutors(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := c.gcByStatusSnapshots(ctx, jobSnapshot, executorSnapshot); err != nil {
+		return nil, nil, err
+	}
+
+	return jobWatchCh, executorReceiver, nil
+}
+
+func (c *DefaultGCCoordinator) gcByStatusSnapshots(
+	ctx context.Context,
+	jobSnapshot JobStatusesSnapshot,
+	executorSnapshot []model.ExecutorID,
+) error {
+	startTime := time.Now()
+	defer func() {
+		duration := time.Since(startTime)
+		log.L().Info("gcByStatusSnapshots finished",
+			zap.Duration("duration", duration))
+	}()
+
+	resources, err := c.metaClient.QueryResources(ctx)
+	if err != nil {
+		return err
+	}
+
+	executorSet := make(map[model.ExecutorID]struct{}, len(executorSnapshot))
+	for _, id := range executorSnapshot {
+		executorSet[id] = struct{}{}
+	}
+
+	var (
+		toGC     []resModel.ResourceID
+		toRemove []resModel.ResourceID
+	)
+	for _, resMeta := range resources {
+		if _, exists := jobSnapshot[resMeta.Job]; !exists {
+			// The resource belongs to a deleted job.
+			toGC = append(toGC, resMeta.ID)
+			continue
+		}
+
+		if _, exists := executorSet[resMeta.Executor]; !exists {
+			// The resource belongs to an offlined executor.
+			toRemove = append(toGC, resMeta.ID)
+			continue
+		}
+	}
+
+	if len(toGC) > 0 {
+		log.L().Info("Adding resources to GC queue",
+			zap.Any("resource-ids", toGC))
+		if err := c.metaClient.SetGCPending(ctx, toGC); err != nil {
+			return err
+		}
+		c.notifier.GCNotify()
+	}
+
+	if len(toRemove) > 0 {
+		log.L().Info("Removing stale resources for offlined executors",
+			zap.Any("resource-ids", toRemove))
+		if _, err := c.metaClient.DeleteResources(ctx, toRemove); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *DefaultGCCoordinator) gcByOfflineJobID(ctx context.Context, jobID string) error {
+	resources, err := c.metaClient.QueryResourcesByJobID(ctx, jobID)
+	if err != nil {
+		return err
+	}
+
+	if len(resources) == 0 {
+		// If there is no resource associated to the job,
+		// we return early.
+		return nil
+	}
+
+	toGC := make([]resModel.ResourceID, 0, len(resources))
+	for _, resMeta := range resources {
+		toGC = append(toGC, resMeta.ID)
+	}
+
+	log.L().Info("Added resources to GC queue",
+		zap.Any("resource-ids", toGC))
+
+	if err := c.metaClient.SetGCPending(ctx, toGC); err != nil {
+		return err
+	}
+
+	c.notifier.GCNotify()
+	return nil
+}
+
+func (c *DefaultGCCoordinator) gcByOfflineExecutorID(ctx context.Context, executorID model.ExecutorID) error {
+	log.L().Info("Cleaning up resources meta for offlined executor",
+		zap.String("executor-id", string(executorID)))
+	return c.metaClient.DeleteResourcesByExecutorID(ctx, string(executorID))
+}
```

Comment:
If we don't need to perform GC operation for offline executor, does this mean all resources are bounded to one executor, what if we support S3 storage as resource in the future.

Reply:
We will adjust this when we implement S3 support. The responsibility of checking whether to perform actual GC should belong to the gcHandler for the resource types.